wcfsetup/install/files/lib/system/package/PackageManifest.class.php

```diff
@@ -0,0 +1,225 @@
+
+/**
+ * Generates a manifest for the given PackageArchive.
+ * 
+ * The manifest is a structured representation of the functional parts
+ * of the archive wherein changes could be relevant to security.
+ * 
+ * @author	Tim Duesterhus
+ * @copyright	2001-2020 WoltLab GmbH
+ * @license	GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @package	WoltLabSuite\Core\System\Package
+ */
+final class PackageManifest {
+	/**
+	 * @var	PackageArchive
+	 */
+	protected $archive;
+	
+	const SUPPORTED_VERSIONS = [1];
+	const CURRENT_VERSION = 1;
+	
+	public function __construct(PackageArchive $archive) {
+		$this->archive = $archive;
+	}
+	
+	/**
+	 * Returns a string containing the manifest version and the manifest's hash.
+	 *
+	 * @param int $version An item in self::SUPPORTED_VERSIONS.
+	 * @return string
+	 */
+	public function getHash($version = self::CURRENT_VERSION) {
+		if ($version === 1) {
+			return $version.'-'.hash('sha256', $this->getManifest($version));
+		}
+		else if (in_array($version, self::SUPPORTED_VERSIONS)) {
+			throw new \LogicException("Unhandled, but supported, manifest version '" . $version . "'");
+		}
+		else {
+			throw new \InvalidArgumentException("Unknown manifest version '" . $version . "'");
+		}
+	}
+	
+	/**
+	 * Returns the archive's manifest.
+	 * 
+	 * @param int $version An element of self::SUPPORTED_VERSIONS.
+	 * @return string
+	 */
+	public function getManifest($version = self::CURRENT_VERSION) {
+		if ($version === 1) {
+			return $this->getManifestV1();
+		}
+		else if (in_array($version, self::SUPPORTED_VERSIONS)) {
+			throw new \LogicException("Unhandled, but supported, manifest version '" . $version . "'");
+		}
+		else {
+			throw new \InvalidArgumentException("Unknown manifest version '" . $version . "'");
+		}
+	}
+	
+	private function getManifestV1() {
+		$requirements = array_map(function ($requirement) {
+			return $requirement['file'];
+		}, $this->archive->getRequirements());
+
+		$optionals = array_map(function ($optional) {
+			if (empty($optional['file'])) {
+				throw new \UnexpectedValueException('Expected to see a file="" attribute for an optional.');
+			}
+			
+			return $optional['file'];
+		}, $this->archive->getOptionals());
+		
+		$includedPackages = array_merge($requirements, $optionals);
+		$ignoredFiles = array_merge($includedPackages, ['package.xml']);
+		
+		return $this->stringifyV1([
+			'manifestVersion' => '1',
+			'identifier' => $this->archive->getPackageInfo('name'),
+			'version' => $this->archive->getPackageInfo('version'),
+			'isApplication' => $this->archive->getPackageInfo('isApplication'),
+			'humanName' => $this->getHumanNames(),
+			'requirements' => $this->getRequirements(),
+			'excludedPackages' => $this->getExcludedPackages(),
+			'files' => $this->getFiles($ignoredFiles),
+			'install' => $this->getInstallInstructions(),
+			'update' => $this->getUpdateInstructions(),
+		]);
+	}
+	
+	private function getHumanNames() {
```

Comment:
```suggestion
	private function getHumanReadableNames() {
```

`getHumanNames()` sounds a bit strange to me? Alternatively, `getDisplayNames()` would be a good candidate too?

Reply:
I guess `getDisplayNames` is the best phrasing here. I'll consistently adjust that everywhere before merging together with the `@since` tag.